# California Mortgage Lending Analysis — Data Understanding

This notebook explores the structure and key variables of the 2024 California HMDA dataset. The goal is to understand the data, evaluate data quality, define the analytical population, and prepare for subsequent data cleaning and exploratory analysis.

**Dataset:** 2024 HMDA Loan-Level Application Register (LAR)  
**Geography:** California  
**Source:** Consumer Financial Protection Bureau (CFPB)

## 1. Setup

Import the Python libraries used for data inspection and analysis.

In [54]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## 2. Dataset Structure

Load the dataset header to inspect the available variables without loading the full dataset into memory.

In [55]:
file_path = "../data/raw/state_CA.csv"

header_df = pd.read_csv(file_path, nrows=0)

print("Number of columns;", len(header_df.columns))


Number of columns; 99


In [56]:
header_df.columns.tolist()

['activity_year',
 'lei',
 'derived_msa-md',
 'state_code',
 'county_code',
 'census_tract',
 'conforming_loan_limit',
 'derived_loan_product_type',
 'derived_dwelling_category',
 'derived_ethnicity',
 'derived_race',
 'derived_sex',
 'action_taken',
 'purchaser_type',
 'preapproval',
 'loan_type',
 'loan_purpose',
 'lien_status',
 'reverse_mortgage',
 'open-end_line_of_credit',
 'business_or_commercial_purpose',
 'loan_amount',
 'loan_to_value_ratio',
 'interest_rate',
 'rate_spread',
 'hoepa_status',
 'total_loan_costs',
 'total_points_and_fees',
 'origination_charges',
 'discount_points',
 'lender_credits',
 'loan_term',
 'prepayment_penalty_term',
 'intro_rate_period',
 'negative_amortization',
 'interest_only_payment',
 'balloon_payment',
 'other_nonamortizing_features',
 'property_value',
 'construction_method',
 'occupancy_type',
 'manufactured_home_secured_property_type',
 'manufactured_home_land_property_interest',
 'total_units',
 'multifamily_affordable_units',
 'income',
 '

## 3. Initial Sample Inspection

Read a small sample of the dataset to inspect record structure, selected mortgage variables, and preliminary data types before loading the full dataset.

In [57]:
sample = pd.read_csv(file_path, nrows=5)

sample

,activity_year,lei,derived_msa-md,state_code,county_code,census_tract,conforming_loan_limit,derived_loan_product_type,derived_dwelling_category,derived_ethnicity,...,denial_reason-2,denial_reason-3,denial_reason-4,tract_population,tract_minority_population_percent,ffiec_msa_md_median_family_income,tract_to_msa_income_percentage,tract_owner_occupied_units,tract_one_to_four_family_homes,tract_median_age_of_housing_units
0,2024,549300GKFNPRWNS0GF29,11244,CA,6059,6059063803,C,Conventional:First Lien,Single Family (1-4 Units):Site-Built,Hispanic or Latino,...,NaN,NaN,NaN,4353,43.03,129000,104,944,1345,50
1,2024,549300GKFNPRWNS0GF29,36084,CA,6013,6013306003,C,FSA/RHS:Subordinate Lien,Single Family (1-4 Units):Site-Built,Hispanic or Latino,...,NaN,NaN,NaN,5615,74.05,155700,66,976,1594,64
2,2024,549300GKFNPRWNS0GF29,31084,CA,6037,6037530802,C,Conventional:First Lien,Single Family (1-4 Units):Site-Built,Not Hispanic or Latino,...,NaN,NaN,NaN,3240,94.72,98200,80,521,882,74
3,2024,549300GKFNPRWNS0GF29,36084,CA,6001,6001451703,C,Conventional:First Lien,Single Family (1-4 Units):Site-Built,Hispanic or Latino,...,NaN,NaN,NaN,3952,41.32,155700,141,1254,1336,45
4,2024,549300GKFNPRWNS0GF29,17020,CA,6007,6007003501,C,Conventional:First Lien,Single Family (1-4 Units):Site-Built,Not Hispanic or Latino,...,NaN,NaN,NaN,3716,46.93,96600,87,707,1290,58


### Key Mortgage Variables

Inspect a focused subset of variables related to geography, loan characteristics, application outcomes, pricing, borrower finances, and property value.

In [58]:
sample[[
    "activity_year",
    "state_code",
    "county_code",
    "loan_type",
    "loan_purpose",
    "loan_amount",
    "action_taken",
    "interest_rate",
    "property_value",
    "income",
    "debt_to_income_ratio"
]]

,activity_year,state_code,county_code,loan_type,loan_purpose,loan_amount,action_taken,interest_rate,property_value,income,debt_to_income_ratio
0,2024,CA,6059,1,1,705000.0,1,6.000,1505000.0,243,41.0
1,2024,CA,6013,4,1,15000.0,1,1.000,485000.0,86,48.0
2,2024,CA,6037,1,1,805000.0,1,6.125,1075000.0,282,46.0
3,2024,CA,6001,1,32,455000.0,1,6.250,995000.0,127,40.0
4,2024,CA,6007,1,1,355000.0,4,NaN,NaN,96,NaN


In [59]:
sample[[
    "activity_year",
    "state_code",
    "county_code",
    "loan_type",
    "loan_purpose",
    "loan_amount",
    "action_taken",
    "interest_rate",
    "property_value",
    "income",
    "debt_to_income_ratio"
]].dtypes

activity_year             int64
state_code                  str
county_code               int64
loan_type                 int64
loan_purpose              int64
loan_amount             float64
action_taken              int64
interest_rate           float64
property_value          float64
income                    int64
debt_to_income_ratio    float64
dtype: object

### Sample Structure and Data Types

Review column data types and non-null counts to identify potential data-quality and type-conversion issues.

In [60]:
sample.info()

<class 'pandas.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 99 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   activity_year                             5 non-null      int64  
 1   lei                                       5 non-null      str    
 2   derived_msa-md                            5 non-null      int64  
 3   state_code                                5 non-null      str    
 4   county_code                               5 non-null      int64  
 5   census_tract                              5 non-null      int64  
 6   conforming_loan_limit                     5 non-null      str    
 7   derived_loan_product_type                 5 non-null      str    
 8   derived_dwelling_category                 5 non-null      str    
 9   derived_ethnicity                         5 non-null      str    
 10  derived_race                              5 non-null 

## 4. Analytical Variable Selection

Select a focused set of variables needed for the mortgage lending analysis. Limiting the dataset to relevant fields reduces memory usage while preserving variables related to geography, application outcomes, loan characteristics, pricing, borrower finances, property characteristics, and denial reasons.

In [61]:
selected_columns = [
    # Identification
    "lei",
    "activity_year",

    # Geography
    "state_code",
    "county_code",
    "census_tract",

    # Application outcome
    "action_taken",

    # Loan characteristics
    "loan_type",
    "loan_purpose",
    "loan_amount",
    "loan_to_value_ratio",
    "loan_term",

    # Pricing
    "interest_rate",
    "rate_spread",
    "total_loan_costs",

    # Borrower financial characteristics
    "income",
    "debt_to_income_ratio",

    # Property
    "property_value",
    "occupancy_type",

    # Denial reasons
    "denial_reason-1",
    "denial_reason-2",
    "denial_reason-3",
    "denial_reason-4"
]

len(selected_columns)

22

In [63]:
missing_columns = [
    col for col in selected_columns
    if col not in header_df.columns
]

missing_columns

[]

## 5. Load Analytical Dataset

Load only the 22 selected variables from the full California HMDA dataset. The Python parsing engine is used for compatibility with the source CSV.

In [65]:
df = pd.read_csv(
    file_path,
    usecols=selected_columns,
    engine="python"
)

In [67]:
df.shape

(1026119, 22)

In [68]:
memory_mb = df.memory_usage(deep=True).sum() / 1024**2
print(f"Dataset memory usage: {memory_mb:.1f} MB")

Dataset memory usage: 537.9 MB


## 6. Data Quality Assessment

Assess the structure, data types, and missingness of the full analytical dataset before defining the population used for lending outcome analysis.

In [69]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1026119 entries, 0 to 1026118
Data columns (total 22 columns):
 #   Column                Non-Null Count    Dtype  
---  ------                --------------    -----  
 0   activity_year         1026119 non-null  int64  
 1   lei                   1026119 non-null  str    
 2   state_code            1026119 non-null  str    
 3   county_code           1019645 non-null  float64
 4   census_tract          1017653 non-null  float64
 5   action_taken          1026119 non-null  int64  
 6   loan_type             1026119 non-null  int64  
 7   loan_purpose          1026119 non-null  int64  
 8   loan_amount           1026119 non-null  float64
 9   loan_to_value_ratio   693086 non-null   str    
 10  interest_rate         647211 non-null   str    
 11  rate_spread           488941 non-null   str    
 12  total_loan_costs      442107 non-null   str    
 13  loan_term             1005146 non-null  str    
 14  property_value        805951 non-null   str  

In [70]:
df.isna().sum().sort_values(ascending=False)

denial_reason-4         1025077
denial_reason-3         1020178
denial_reason-2          991458
total_loan_costs         584012
rate_spread              537178
interest_rate            378908
debt_to_income_ratio     362074
loan_to_value_ratio      333033
property_value           220168
income                   140957
loan_term                 20973
census_tract               8466
county_code                6474
lei                           0
action_taken                  0
state_code                    0
activity_year                 0
loan_amount                   0
loan_purpose                  0
loan_type                     0
occupancy_type                0
denial_reason-1               0
dtype: int64

## 7. Application Outcome Distribution

Examine the distribution of HMDA action codes to understand how applications were resolved and to define the appropriate population for approval and denial analysis.

In [71]:
df["action_taken"].value_counts().sort_index()

action_taken
1    511270
2     36102
3    180539
4    142155
5     55626
6     96647
7       584
8      3196
Name: count, dtype: int64

In [72]:
df["action_taken"].value_counts(normalize=True).sort_index() * 100

action_taken
1    49.825605
2     3.518305
3    17.594353
4    13.853656
5     5.421009
6     9.418693
7     0.056913
8     0.311465
Name: proportion, dtype: float64

In [73]:
action_labels = {
    1: "Loan Originated",
    2: "Approved, Not Accepted",
    3: "Denied",
    4: "Withdrawn",
    5: "Closed for Incompleteness",
    6: "Purchased Loan",
    7: "Preapproval Denied",
    8: "Preapproval Approved, Not Accepted"
}

In [74]:
df["action_label"] = df["action_taken"].map(action_labels)

In [75]:
df[["action_taken", "action_label"]].head(10)

,action_taken,action_label
0,1,Loan Originated
1,1,Loan Originated
2,1,Loan Originated
3,1,Loan Originated
4,4,Withdrawn
5,1,Loan Originated
6,1,Loan Originated
7,1,Loan Originated
8,4,Withdrawn
9,1,Loan Originated


In [76]:
df["action_label"].value_counts()

action_label
Loan Originated                       511270
Denied                                180539
Withdrawn                             142155
Purchased Loan                         96647
Closed for Incompleteness              55626
Approved, Not Accepted                 36102
Preapproval Approved, Not Accepted      3196
Preapproval Denied                       584
Name: count, dtype: int64

## 8. Define Decisioned Applications

For approval-rate analysis, restrict the population to applications with a clear lending decision:

- **Approved:** Loan originated (`1`) or approved but not accepted (`2`)
- **Denied:** Application denied (`3`)

Withdrawn, incomplete, purchased, and preapproval records are excluded because they do not represent comparable final underwriting decisions in this analysis.

In [77]:
decisioned_df = df[
  df["action_taken"].isin([1, 2, 3])
].copy()

In [78]:
decisioned_df.shape

(727911, 23)

In [79]:
decisioned_df["approved"] = (
    decisioned_df["action_taken"].isin([1, 2])
).astype(int)

In [80]:
decisioned_df[
    ["action_taken", "action_label", "approved"]
].head(10)

,action_taken,action_label,approved
0,1,Loan Originated,1
1,1,Loan Originated,1
2,1,Loan Originated,1
3,1,Loan Originated,1
5,1,Loan Originated,1
6,1,Loan Originated,1
7,1,Loan Originated,1
9,1,Loan Originated,1
10,1,Loan Originated,1
11,1,Loan Originated,1


## 9. Approval Rates by Loan Type

Compare approval outcomes across major mortgage loan types within the decisioned application population. Both application volume and approval rate are reported because percentage differences should be interpreted in the context of sample size.

In [81]:
overall_approval_rate = decisioned_df["approved"].mean()

print(f"Overall approval rate: {overall_approval_rate:.2%}")

Overall approval rate: 75.20%


In [82]:
loan_type_labels = {
    1: "Conventional",
    2: "FHA",
    3: "VA",
    4: "USDA/RHS"
}

decisioned_df["loan_type_label"] = (
    decisioned_df["loan_type"].map(loan_type_labels)
)

## Loan Type Approval Summary

In [85]:
loan_type_summary = (
    decisioned_df
    .groupby("loan_type_label")["approved"]
    .agg(["count", "sum", "mean"])
)

In [86]:
loan_type_summary = loan_type_summary.rename(
    columns={
        "count": "applications",
        "sum": "approved",
        "mean": "approval_rate"
    }
)

In [87]:
loan_type_summary["denied"] = (
    loan_type_summary["applications"]
    - loan_type_summary["approved"]
)

In [90]:
loan_type_summary["approval_rate_pct"] = (
    loan_type_summary["approval_rate"] * 100
)

loan_type_summary = loan_type_summary.sort_values(
    "approval_rate_pct",
    ascending=False
)

loan_type_summary[
    ["applications", "approved", "denied", "approval_rate_pct"]
].round(2)

,applications,approved,denied,approval_rate_pct
loan_type_label,,,,
VA,31840,26367,5473,82.81
USDA/RHS,423,342,81,80.85
FHA,78540,61978,16562,78.91
Conventional,617108,458685,158423,74.33


### Validation

In [91]:
assert (
    loan_type_summary["approved"]
    + loan_type_summary["denied"]
    == loan_type_summary["applications"]
).all()

### Interpretation

Among decisioned applications, the overall approval rate is approximately **75.2%**. Approval rates vary across loan types, ranging from **74.3% for conventional loans** to **82.8% for VA loans**.

These descriptive differences should not be interpreted as evidence that one loan type is inherently easier to obtain. The borrower populations and application characteristics differ across loan programs, including financial profiles, loan amounts, property characteristics, and loan purposes. VA loans also serve an eligibility-restricted population, while the USDA/RHS group contains a much smaller sample than the other loan types.

Therefore, these results show an **association between loan type and observed approval outcomes**, not a causal effect of loan type on approval probability.

# 10. Approval Rates by Loan Purpose

Compare approval outcomes across mortgage loan purposes within the decisioned application population.

Before comparing approval rates, examine the distribution of loan-purpose categories and map the HMDA codes to descriptive labels. Application volume is considered alongside approval rates because observed differences may reflect differences in borrower and application characteristics.

In [110]:
decisioned_df["loan_purpose"].value_counts().sort_index()

loan_purpose
1     310494
2      96339
4     107230
5        509
31     88318
32    125021
Name: count, dtype: int64

In [111]:
loan_purpose_pct = (
    decisioned_df["loan_purpose"]
    .value_counts(normalize=True)
    .sort_index()
    * 100
)

loan_purpose_pct

loan_purpose
1     42.655489
2     13.234997
4     14.731197
5      0.069926
31    12.133077
32    17.175314
Name: proportion, dtype: float64

## Loan Purpose Code Mapping

Map HMDA loan-purpose codes to descriptive labels before performing group-level analysis.

In [112]:
loan_purpose_labels = {
    1: "Home Purchase",
    2: "Home Improvement",
    4: "Other Purpose",
    5: "Not Applicable",
    31: "Refinancing",
    32: "Cash-out Refinancing"
}

In [113]:
decisioned_df["loan_purpose_label"] = (
    decisioned_df["loan_purpose"].map(loan_purpose_labels)
)

In [114]:
decisioned_df[
    ["loan_purpose", "loan_purpose_label"]
].drop_duplicates().sort_values("loan_purpose")

,loan_purpose,loan_purpose_label
0,1,Home Purchase
96,2,Home Improvement
38,4,Other Purpose
56513,5,Not Applicable
5,31,Refinancing
3,32,Cash-out Refinancing


## Loan Purpose Approval Summary

Summarize application volume, approvals, denials, and approval rates for each loan-purpose category.

In [115]:
loan_purpose_summary = (
    decisioned_df
    .groupby("loan_purpose_label")["approved"]
    .agg(["count", "sum", "mean"])
)

In [116]:
loan_purpose_summary = loan_purpose_summary.rename(
    columns={
        "count": "applications",
        "sum": "approved",
        "mean": "approval_rate"
    }
)

In [117]:
loan_purpose_summary["denied"] = (
    loan_purpose_summary["applications"]
    - loan_purpose_summary["approved"]
)

In [118]:
loan_purpose_summary["approval_rate_pct"] = (
    loan_purpose_summary["approval_rate"] * 100
)

In [119]:
loan_purpose_summary = loan_purpose_summary.sort_values(
    "approval_rate_pct",
    ascending=False
)

In [121]:
loan_purpose_summary[
    [
        "applications",
        "approved",
        "denied",
        "approval_rate_pct"
    ]
].round(2)

,applications,approved,denied,approval_rate_pct
loan_purpose_label,,,,
Not Applicable,509,454,55,89.19
Home Purchase,310494,274579,35915,88.43
Refinancing,88318,67353,20965,76.26
Cash-out Refinancing,125021,87185,37836,69.74
Home Improvement,96339,58235,38104,60.45
Other Purpose,107230,59566,47664,55.55


### Validation

Confirm that approved and denied applications reconcile to the total decisioned applications within each loan-purpose category.

In [123]:
assert (
    loan_purpose_summary["approved"]
    + loan_purpose_summary["denied"]
    == loan_purpose_summary["applications"]
).all()

## Interpretation

Among loan purposes with clearly defined categories, **Home Purchase** has the highest observed approval rate at approximately **88.43%**, while **Other Purpose** has the lowest observed approval rate at approximately **55.55%**.

Traditional refinancing applications show an observed approval rate of approximately **76.26%**, compared with **69.74%** for Cash-out Refinancing.

The **Not Applicable** category has the highest numerical approval rate at approximately **89.19%**, but it contains only **509 decisioned applications**, substantially fewer than the other categories. Its approval rate should therefore be interpreted cautiously.

These descriptive differences should not be interpreted as evidence that one loan purpose is inherently easier to obtain approval for than another. Borrower and application characteristics may differ systematically across loan-purpose groups, including income, debt-to-income ratio, loan-to-value ratio, loan amount, property characteristics, and loan type.

Therefore, the results indicate an **association between loan purpose and observed approval outcomes**, rather than a causal effect of loan purpose on approval probability.